### PR TITLE
Remove Python object reference cycles from base and server classes

### PR DESCRIPTION
**Pull request by toonen**
#### Describe Bug or Feature

In the PBS Python server classes (ex: `_job`) used during hook invocation, some class level attributes, such as the `_attributes_hook_set` and those set to a `PbsAttributeDescriptor`, exist as dictionaries whose keys are a references to objects instantiated from those classes.  This creates a reference cycle where an object refers to the class (in `__class__`) and the class level dictionaries hold one or more references to the object.  The following diagram is an example of such a cycle.

```
_job._attributes_hook_set[_job()] = {'id': None}
^    ^                    ^         ^
|    |                    |         Value cached
|    |                    Object Instance of class _job, increments reference to that object.
|    |                    This also stores the object reference on the class which persists until python is restarted.
|    Class attribute 
Class
```

These cycles prevent the Python garbage collector from being able to free the event object and all objects it directly or indirectly references until such time as the class attributes are deleted.  Over time this results in a large accumulation of objects that are no longer needed.  The current solution is to periodically restart the Python interpreter.  When the interpreter is shutdown, the class attributes are deleted, breaking the reference cycles and allowing the objects to be deleted as well.  Unfortunately, periodically restarting the interpreter has side effects, including crashing the server when some Python extension modules (shared libraries) are used.

#### Describe Your Change

To prevent these reference cycles and allow the objects to be freed by the garbage collector, we replaced the existing dictionaries with instances of a `WeakKeyDictionary` provided by the weakref module.  The object references in the dictionaries will persist until such time as the last strong reference to the object is removed.  At that time, the object will be freed and any references to it in the dictionaries will automatically be removed.

With these changes, we are able to disable the Python interpreter restart and observe that the amount of memory used by the server,  as well as the number of objects being tracked by the Python garbage collector, remains relatively constant over time.

#### **Implementation Update**:

Unfortunately, the use of WeakKeyDictionary objects failed to capture updates made to attributes and resources.  The short answer as to why is that a WeakKeyDictionary is a MutableMapping, as is a dict, but it is not a dict.  The upshot is that the C API PyDict functions do not work with WeakKeyDictionary objects.  We headed down the path of using the PyMapping and PyObject functions instead; however, during the process, we realized the real issue was that the mappings of class instances (objects) as keys in class-level dictionaries were unnecessary.  As such, we moved the attribute and resource tracking data structures into the objects, which resulted in somewhat simpler code.

### CODE REVIEW

**Review by bayucan — COMMENTED** [review]
This is a nice discovery! Thanks. It looks reasonable. I would like evidence of testing, and in particular, please run the following PTL tests, which are under tests/functional. These tests make use of Python hooks:
[test.list2.txt](https://github.com/openpbs/openpbs/files/7565670/test.list2.txt)

**Comment by bayucan** [comment]
@toonen : anything else to add to this PR (i.e. test output) or is it good to go? I've seen the test results and they looked fine.

**Comment by toonen** [comment]
@bayucan I believe it is ready to be merged.  @ericpershey included the test results in a previous comment.

**Review by bayucan — APPROVED** [review]
Looks good.

**Comment by bayucan** [comment]
@toonen : ok, I approved this. Will wait for another reviewer before merging. @openpbs/openpbs-maintainers : this needs another erviewer.

File: test/tests/functional/pbs_acct_log.py
Comment (bhroam) [inline test/tests/functional/pbs_acct_log.py:673]: Same here, I don't think extend=x does anything on a submit.

**Review by bhroam — CHANGES_REQUESTED** [review]

File: test/tests/functional/pbs_acct_log.py
Comment (ericpershey) [reply test/tests/functional/pbs_acct_log.py:471]: I was doing deeper tests with types.  Will change it to 42.

**Comment by ericpershey** [comment]
@bhroam Thanks for all of the feedback.  Commit 28167703df239f1bd8944df6609480c956e2ab58 addresses the various issues you raised.

**Review by bhroam — APPROVED** [review]
LGTM

**Comment by bhroam** [comment]
@bayucan do you want to take another look, or should this be merged?

**Comment by bayucan** [comment]
@bhroam : still looks good to me. I'll merge.

**Comment by bayucan** [comment]
Merged #2501 into master.

**Pull request merged**
